packages/jupyterlab-manager/src/semverregistry.ts

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { maxSatisfying } from 'semver';
+
+import { ArrayExt } from '@lumino/algorithm';
+
+import { PromiseDelegate } from '@lumino/coreutils';
+
+import { Debouncer } from '@lumino/polling';
+
+const TIMEOUT = 300;
+
+interface IRequest<T> {
+  name: string;
+  semVer: string;
+  result: PromiseDelegate<T | undefined>;
+  timestamp: number | null;
+}
+
+/**
+ * A registry using semver ranges to retrieve values.
+ *
+ * Each object in the registry is associated with a name and a semantic
+ * version. Objects can be retrieved by giving the name and a semantic version
+ * range, and the maximum satisfying version of the object will be returned.
+ *
+ * Object retrieval may be delayed for a short timeout if no matching object
+ * can be found. This is to allow a short window for objects to be registered
+ * before get requests are resolved as unsatisfied. Practically, this means
+ * that initial object registration can happen concurrent with the first
+ * requests.
+ *
+ */
+export class SemVerRegistry<T> {
+  /**
+   * Register a object with a given name and version.
+   */
+  register(options: SemVerRegistry.IRegistryData<T>): void {
+    const { name, version, object } = options;
+    let versions = this._registry.get(name);
+    if (versions === undefined) {
+      versions = new Map<string, T>();
+      this._registry.set(name, versions);
+    }
+
+    if (versions.has(version)) {
+      throw new Error(`Version ${version} of ${name} already registered.`);
+    }
+    versions.set(version, object);
+
+    // Some pending requests may now be satisfied, so schedule a retry.
+    if (this._pendingRequests.length > 0) {
+      this._debounce.invoke();
+    }
+  }
+
+  /**
+   * Get an object for a name and semVer range.
+   *
+   * @returns a promise resolving to the object with the maximum version
+   * satisfying the version range if one is registered, or undefined if one is
+   * not registered.
+   *
+   * #### Notes
+   * Resolution may be delayed for a short timeout to allow for some delay in
+   * registering the object.
+   */
+  async get(
+    name: string,
+    semVer: string,
+    isKernelRestore: boolean
+  ): Promise<T | undefined> {
+    const mod = this._getSync(name, semVer);
+    if (mod || this._already_retried.indexOf(name + ':' + semVer)) {
+      return mod;
+    }
+    this._already_retried.push(name + ':' + semVer);
+    const result = new PromiseDelegate<T | undefined>();
+    let timestamp = Date.now() + TIMEOUT;
+    if (isKernelRestore) timestamp += TIMEOUT * 4;
+    this._pendingRequests.push({ name, semVer, result, timestamp });
+    this._debounce.invoke();
```

Comment:
If `this._throttle.invoke();` doesn't work because we are invoking from the invoked state, we can always invoke it asynchronously: `setTimeout(()=>this._debounce.invoke, 0)` or something.

Reply:
I filed an issue at https://github.com/jupyterlab/lumino/issues/47